routing/turns_generator.cpp

```diff
@@ -1151,6 +1207,13 @@ void GetTurnDirection(IRoutingResult const & result, size_t outgoingSegmentIndex
     return;
   }
 
+  if (!turn.m_keepAnyway && IsGoStraightOrSlightTurn(turn.m_turn) && nodes.candidates.size() != 1 &&
```

Comment:
1. Хороший вопрос.
Просто скопировал как в KeepTurnByHighwayClass.
Я не вижу проблем с `nodes.candidates.size() == 1`.
2. Предположительно, не может.
@vng

Reply:
Не может, тут все хорошо и да по аналогии с предидущим сравнением.
Эквивалентно nodes.candidates.size() > 1, проверяем когда больше одной альтернативы

Reply:
Тогда надо не копипастить, а сделать так:
```C++
  if (!turn.m_keepAnyway && IsGoStraightOrSlightTurn(turn.m_turn) && nodes.candidates.size() != 1)
  {
    if (!KeepTurnByHighwayClass(nodes, turnInfo, numMwmIds, turn.m_turn) ||
        !KeepTurnByAlignedAlternatives(nodes, turnInfo, numMwmIds))
    {
      turn.m_turn = CarDirection::None;
      return;
    }
  }
```

Reply:
вопросы к @vng :
```
Note 2. If there's only one exit from this junction (nodes.candidates.size() != 1)
  // this turn should be kept.
```
- если выезд только один, то nodes.candidates.size() == 1, а не наоборот?
- если выезд только один - то почему именно в этом случае нужно оставить поворот? Если этого условия не будет, то оба KeepTurnBy... превратят SlightTurn в None. По идее, он и так уже должен быть таким к этому моменту.

Reply:
Я эти комменты не писал :) но предполагаю что:
- Для одного exit когда nodes.candidates.size() == 1, мы НЕ заходим в KeepTurnByHighwayClass и turn точно сохраняется. Можете переписать этот коммент на понятный :)
- Вероятно он остается, потому что вычисляется по какому-то другому жесткому условию раньше (например это может быть пересадка на паром), тут не скажу точно. Пусть так и остается, пока не увидим явный косяк

Reply:
Меняю, если нет возражений, на
```
  // Note 2. If there's only one exit from this junction (nodes.candidates.size() == 1)
  // this turn should be kept (because if was kept by previous logic as an exception).
```

Reply:
После перехода на
```
if (!KeepTurnByHighwayClass(nodes, turnInfo, numMwmIds, turn.m_turn) ||
        !KeepTurnByAlignedAlternatives(nodes, turnInfo, numMwmIds))
```
Стала понятна немного перевернутая логика в названиях функций.
Исходя из названия, должно быть так
```
if (! (KeepTurnByHighwayClass(nodes, turnInfo, numMwmIds, turn.m_turn) ||
        KeepTurnByAlignedAlternatives(nodes, turnInfo, numMwmIds)))
```
те если ни одной причины нет, чтобы сохранить поворот - то мы его удаляем.
а у нас получается, что если причины_1 нет - удаляем, если причины 2 нет - удаляем.

Предлагаю инвертировать функции на DiscardTurnByHighwayClass и DiscardTurnByAlignedAlternatives.
Тогда получится:
```
if (DiscardTurnByHighwayClass(nodes, turnInfo, numMwmIds, turn.m_turn) ||
        DiscardTurnByAlignedAlternatives(nodes, turnInfo, numMwmIds))
```

Reply:
согласен переименовать

Reply:
в процессе переименования обратил внимание на
```
bool KeepTurnByHighwayClass()
  ...
  if (!GetTurnHighwayClasses(possibleTurns, turnInfo, numMwmIds, turnHighwayClasses))
    return false;
```

Reply:
Исходя из описания GetTurnHighwayClasses
If it's possible fills |turnHighwayClasses| and returns true, and returns false otherwise.
это выглядит ошибкой: если данные не получили - то поворот не нужен. должно быть наоборот?

Reply:
Есть подозрение, что 
GetTurnHighwayClasses
не совсем соответствует описанию в отношении возвращаемого значения.

Reply:
Оставил старую логику (инвертировав true на false), добавив комент:
```
bool DiscardTurnByHighwayClass(TurnCandidates const & possibleTurns, TurnInfo const & turnInfo,
                               NumMwmIds const & numMwmIds, CarDirection carDirection)
...
  // Looks like a bug. If no info received turn should not be discarded
  if (!GetTurnHighwayClasses(possibleTurns, turnInfo, numMwmIds, turnHighwayClasses))
    return true;
```